src/MongoDB.Driver.Core/Core/Connections/TcpStreamFactory.cs

```diff
@@ -143,43 +141,54 @@ private void Connect(Socket socket, EndPoint endPoint, CancellationToken cancell
                     {
                         socket.Connect(endPoint);
                     }
-                    connected = true;
-                    return;
+                    ChangeState(2); // note: might not actually go to state 2 if already in state 3 or 4
                 }
                 catch
                 {
-                    if (!cancelled && !timedOut)
+                    if (state == 1)
                     {
                         try { socket.Dispose(); } catch { }
                         throw;
                     }
+                    // a timeout or cancellation exception will be thrown instead below
                 }
-            }
 
-            try { socket.Dispose(); } catch { }
+                switch (state)
+                {
```

Comment:
Why we cannot have here the same lines as in the previous PR: 

                if (state == 3) { throw new TimeoutException(); }
                if (state == 4) { throw new OperationCanceledException(); }

From my point of view, it's the same logic (except the _default_ case which will never happen)

Reply:
We don't want to change the timeout exception message do we? 

Other than that sure I can make that change.

Reply:
We can add the message inside brackets, my concern that here we can have a very similar solution as in the previous PR (2884). And I think whether we have or don't have the error message for timeout exception doesn't make any real difference between them